lamagraph-core/src/Core/Reducer.hs

```diff
@@ -0,0 +1,95 @@
+module Core.Reducer where
+
+import Clash.Prelude
+import Control.Lens (makeLenses, set, (^.))
+import Core.MemoryManager
+import Core.Node
+
+type NumOfNodesToStore = Unsigned 3
+type NumOfEdgesToStore = Unsigned 3
+
+{- | Regulates which external (interface) `Ports` belonged to which `Node` (and in which place) before the reduction.
+  This is necessary in order to be able to coordinate the interface in the reducer
+-}
+data OldId portsNumber = LeftLoadedNode (IdOfPort portsNumber) | RightLoadedNode (IdOfPort portsNumber)
+
+-- | Result of abstract reduction rule
+data ReduceRuleResult (nodesNumber :: Nat) (edgesNumber :: Nat) (portsNumber :: Nat) = ReduceRuleResult
+  { _edges :: Vec edgesNumber (Maybe (OldId portsNumber, OldId portsNumber))
+  , _nodes ::
+      Vec nodesNumber (Maybe (LocalNode portsNumber, Vec portsNumber (Maybe (OldId portsNumber, IdOfPort portsNumber))))
+  }
+
+$(makeLenses ''ReduceRuleResult)
+
+{- | Coordinates interface `Port` in `LocalNode`.
+I.e. it connects the "hanging" ports from the reduction rule with the `LoadedNode` from the real net
+-}
+putInterfacesNodeToGlobalNet ::
+  forall portsNumber.
+  (KnownNat portsNumber) =>
+  LoadedNode portsNumber ->
+  LoadedNode portsNumber ->
+  (LocalNode portsNumber, Vec portsNumber (Maybe (OldId portsNumber, IdOfPort portsNumber))) ->
+  LocalNode portsNumber
+putInterfacesNodeToGlobalNet leftLNode rightLNode (localNode, interfacePortsInfo) = foldl replaceOnePort localNode interfacePortsInfo
+ where
+  replaceOnePort :: LocalNode portsNumber -> Maybe (OldId portsNumber, IdOfPort portsNumber) -> LocalNode portsNumber
+  replaceOnePort ln maybePortInfo =
+    case maybePortInfo of
```

Comment:
I'd suggest using LambdaCase here, mostly to reduce indentation, but also to get rid of unused identifier.
IMO LambdaCase must be used throughout project instead of `... arg = case arg of` when `arg` isn't used in the body. Also, `\cases` from the same extension look useful